File: LeetCode/Math/JS/Leetcode1175-prime-arrangements.js
var numPrimeArrangements = function (n) {
    let [st, primes] = [new Array(n + 1).fill(false), []];
    const P = parseInt(1e9 + 7);
    for (let i = 2; i <= n; i++) {
        if (!st[i]) primes.push(i);
        for (let j = 0; primes[j] * i <= n; j++) {
            st[primes[j] * i] = true;
            if (i % primes[j] == 0) break;
        }
    }
    let cnt = primes.length;
    let res = 1;
    let m = n - cnt;
    while (cnt > 0) {
        res = res % P;
        res *= cnt;
        cnt--;
    }
    while (m > 0) {
        res = res % P;
        res *= m;
        m--;
    }
    return res % P;
};

File: LeetCode/Math/Golang/Leetcode1175-prime-arrangements.go
package sbo

func numPrimeArrangements(n int) int {
    primes,st := []int{},make([]bool,n + 1)
    for i := 2;i <= n;i++ {
        if !st[i] {
            primes = append(primes,i)
        }
        for j := 0;primes[j] <= n / i;j++ {
            st[primes[j] * i] = true
            if i % primes[j] == 0 {
                break
            }
        }
    }
    cnt := len(primes)
    const P = 1e9 + 7
    fact := func(n int) int {
        res := 1
        for i := 1;i <= n;i++ {
            res = res * i % P
        }
        return res
    }
    return fact(n - cnt) * fact(cnt) % P
}